File: js/streams/client/event.ts
import {
    ErroneousEvent as ErroneousEventPb,
    Event,
} from "@fraym/proto/dist/index.freym.streams.management";

export interface SubscriptionEvent extends BaseEvent {
    topic: string;
    raisedAt: Date;
    orderSerial?: number;
    deploymentId?: number;
    retryNumber?: number;
}

export interface PublishEvent extends BaseEvent {
    broadcast?: boolean;
}

export interface BaseEvent {
    id: string;
    payload: Record<string, EventData>;
    tenantId: string;
    type?: string;
    stream?: string;
    correlationId?: string;
    causationId?: string;
    reason?: string;
    userId?: string;
}

/* eslint-disable-next-line @typescript-eslint/no-explicit-any */
export type EventData = any | GdprEventData;

export interface GdprEventData {
    id?: string;
    /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
    value: any;
    /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
    gdprDefault: any;
    isInvalidated: boolean;
}

export const isGdprEventData = (value: EventData): value is GdprEventData => {
    return (
        value &&
        typeof value === "object" &&
        Object.keys(value).length == 2 &&
        Object.prototype.hasOwnProperty.call(value, "value") &&
        Object.prototype.hasOwnProperty.call(value, "gdprDefault")
    );
};

export type HandlerFunc = (event: SubscriptionEvent) => Promise<void>;

export const getSubscriptionEvent = (event: Event): SubscriptionEvent => {
    const payload: Record<string, EventData> = {};

    for (const key in event.payload) {
        const data = event.payload[key];

        if (data === undefined) {
            continue;
        }

        if (data.gdpr) {
            payload[key] = {
                id: data.gdpr.id,
                value: JSON.parse(data.value),
                gdprDefault: data.gdpr.default ? JSON.parse(data.gdpr.default) : "",
                isInvalidated: data.gdpr.isInvalidated,
            };
        } else {
            payload[key] = JSON.parse(data.value);
        }
    }

    return {
        id: event.id,
        topic: event.topic,
        tenantId: event.tenantId,
        payload,
        raisedAt: new Date(parseInt(event.raisedAt.slice(0, -6))),
        stream: event.stream || undefined,
        type: event.type || undefined,
        causationId: event.metadata ? event.metadata.causationId : undefined,
        correlationId: event.metadata ? event.metadata.correlationId : undefined,
        reason: event.reason || undefined,
        orderSerial: event.metadata ? parseInt(event.metadata.orderSerial) : undefined,
        deploymentId:
            event.metadata && event.metadata.deploymentId
                ? parseInt(event.metadata.deploymentId)
                : undefined,
        userId: event.metadata ? event.metadata.userId || undefined : undefined,
        retryNumber: event.metadata ? parseInt(event.metadata.retryNumber) : undefined,
    };
};

export interface ErroneousEvent {
    event: SubscriptionEvent;
    consumerGroup: string;
    error: string;
}

export const getErroneousEvent = (event: ErroneousEventPb): ErroneousEvent | null => {
    if (!event.event) {
        return null;
    }

    return {
        event: getSubscriptionEvent(event.event),
        consumerGroup: event.consumerGroup,
        error: event.error,
    };
};

File: go/streams/internal/util/event.go
package util

import (
	"encoding/json"
	"time"

	"github.com/fraym/freym-api/go/proto/streams/managementpb"
	"github.com/fraym/freym-api/go/streams/domain/dto"
)

func SubscriptionEventsFromPb(data []*managementpb.Event) ([]*dto.SubscriptionEvent, error) {
	var events []*dto.SubscriptionEvent

	for _, eventData := range data {
		event, err := SubscriptionEventFromPb(eventData)
		if err != nil {
			return nil, err
		}

		events = append(events, event)
	}

	return events, nil
}

func SubscriptionEventFromPb(data *managementpb.Event) (*dto.SubscriptionEvent, error) {
	payload := map[string]dto.EventPayload{}

	for key, value := range data.GetPayload() {
		var valueData any
		err := json.Unmarshal([]byte(value.GetValue()), &valueData)
		if err != nil {
			return nil, err
		}

		var gdprPayload *dto.GdprEventPayload

		gdprData := value.GetGdpr()
		if gdprData != nil {
			var defaultValue any
			err := json.Unmarshal([]byte(gdprData.GetDefault()), &defaultValue)
			if err != nil {
				return nil, err
			}

			gdprPayload = &dto.GdprEventPayload{
				Id:          gdprData.GetId(),
				Default:     defaultValue,
				Invalidated: gdprData.GetIsInvalidated(),
			}
		}

		payload[key] = dto.EventPayload{
			Value: valueData,
			Gdpr:  gdprPayload,
		}
	}

	var (
		correlationId string
		causationId   string
		orderSerial   int64
		userId        string
		deploymentId  int64
		retryNumber   int64
	)

	if meta := data.GetMetadata(); meta != nil {
		correlationId = meta.GetCorrelationId()
		causationId = meta.GetCausationId()
		orderSerial = meta.GetOrderSerial()
		userId = meta.GetUserId()
		deploymentId = meta.GetDeploymentId()
		retryNumber = meta.GetRetryNumber()
	}

	return &dto.SubscriptionEvent{
		Id:            data.GetId(),
		TenantId:      data.GetTenantId(),
		Stream:        data.GetStream(),
		Type:          data.GetType(),
		CorrelationId: correlationId,
		CausationId:   causationId,
		Reason:        data.GetReason(),
		Payload:       payload,
		Topic:         data.GetTopic(),
		RaisedAt:      time.Unix(0, data.GetRaisedAt()),
		OrderSerial:   orderSerial,
		UserId:        userId,
		DeploymentId:  deploymentId,
		RetryNumber:   retryNumber,
	}, nil
}

func ErroneousEventsFromPb(data []*managementpb.ErroneousEvent) ([]*dto.ErroneousEvent, error) {
	var events []*dto.ErroneousEvent

	for _, eventData := range data {
		event, err := SubscriptionEventFromPb(eventData.GetEvent())
		if err != nil {
			return nil, err
		}

		events = append(events, &dto.ErroneousEvent{
			Event:         event,
			ConsumerGroup: eventData.GetConsumerGroup(),
			Error:         eventData.GetError(),
		})
	}

	return events, nil
}
